clang/runtime/dpct-rt/include/dpct/bindless_images.hpp

```diff
@@ -117,6 +137,341 @@ class image_mem_wrapper {
+
+      _img_mem_handle_ptr =
+        new sycl::ext::oneapi::experimental::image_mem_handle;
+      *_img_mem_handle_ptr =
+        sycl::ext::oneapi::experimental::map_external_image_memory(
+          get_interop_mem_handle(), _img_desc, q);
+      
+      _img_mem_wrapper_ptr = new image_mem_wrapper(this);
+    }
+  }
+
+  image_mem_wrapper* get_mapped_mipmapped_array() {
+    assert(!_res_is_buffer &&
+        "Buffer resouce cannot be accessed as an array!");
+    assert(_img_mem_wrapper_ptr && "Resouce is not mapped!");
+
+    return _img_mem_wrapper_ptr;
+  }
+
+  image_mem_wrapper* get_sub_resource_mapped_array(
+      unsigned array_index, unsigned mip_level) {
+    assert(!_res_is_buffer &&
+        "Buffer resouce cannot be accessed as an array!");
+    assert(array_index == 0 &&
+        "Retrieving a particular layer from image array is not yet supported!");
+    assert(array_index < _img_desc.array_size &&
+        "Array index requested exceeds total array size in resource!");
+    assert(mip_level < _img_desc.num_levels &&
+        "Mip level requested exceeds total num of levels in resource!");
+    assert(_img_mem_wrapper_ptr && "Resouce is not mapped!");
+
+    return (mip_level == 0)?
+          _img_mem_wrapper_ptr: 
+          _img_mem_wrapper_ptr->get_mip_level(mip_level);
+  }
+
+  void unmap_resource(sycl::queue q) {
+    if (_res_is_buffer) {
+      assert(_buf_ptr && "Resource is not mapped!");
+
+      _buf_ptr = nullptr;
+
+      throw std::runtime_error(
+          "Unmapping a buffer resoure is not yet supported!");
+    }
+    else {
+      assert(_img_mem_handle_ptr && "Resource is not mapped!");
+
+      delete _img_mem_wrapper_ptr;
+      _img_mem_wrapper_ptr = nullptr;
+      _img_mem_handle_ptr = nullptr;
+    }
+  }
+
+#ifdef _WIN32
+  // Helper function to query the properties of DX11 resource
+  void query_res_info(ID3D11Resource *resource) {
+    unsigned res_width = 1;
+    unsigned res_height = 1;
+    unsigned res_depth = 1;
+    unsigned res_num_channels = 1;
+    unsigned res_num_levels = 1;
+    unsigned res_arr_size = 1;
+    unsigned res_fomat_size_bytes = 0;
+
+    sycl::image_channel_type img_ch_type =
+      sycl::image_channel_type::unsigned_int8;
+
+    // Get the dimension info of DX11 resource
+    D3D11_RESOURCE_DIMENSION dimension;
+    resource->GetType(&dimension);
+
+    switch (dimension) {
+      case D3D11_RESOURCE_DIMENSION_BUFFER: {
+        _res_is_buffer = true;
+        
+        D3D11_BUFFER_DESC desc;
+        ((ID3D11Buffer*)resource)->GetDesc(&desc);
+
+        res_width = desc.ByteWidth;
+
+        _res_size_bytes = desc.ByteWidth;
+        break;
+      }
+      case D3D11_RESOURCE_DIMENSION_TEXTURE1D: {
+        D3D11_TEXTURE1D_DESC desc;
+        ((ID3D11Texture1D*)resource)->GetDesc(&desc);
+
+        res_width = desc.Width;
+        res_arr_size = desc.ArraySize;
+        res_num_levels = (desc.MipLevels? desc.MipLevels: 1);
+
+        get_img_ch_info(
```

Comment:
Can create a dpct::image_channel object to get the total bytes. Then need not add the `get_img_ch_info` function.